rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+
+.add_systems([run, jump].after(InputSet::ReadInput))
+```
+
+- looks pretty
+- no limit to number of elements
+- doesn't work (different functions are different types, arrays must be homogoneous)
+
+#### Tuples
+
+```rust
+.add_systems((
+        compute_attack, 
+        compute_defense,
+        check_for_crits,
+        compute_damage,
+        deal_damage,
+        check_for_death
+    )
+    .chain()
+    .in_set(GameSet::Combat)
+)
+
+.add_systems((run, jump).after(InputSet::ReadInput))
+```
+
+- looks pretty
+- limited to 12 elements
+- increases compile times
+- relies on type system magic
+  - requires invocation of the `all_tuples` macro to generate blanket impls
+  - (kinda cursed, but already used for system params, systems, and bundles)
+
+#### `vec!`-like macros
+
+This strategy is used for the `vec![1,2,3]` macro in the standard library.
+
+```rust
+.add_systems(systems![
+      compute_attack, 
+      compute_defense,
+      check_for_crits,
+      compute_damage,
+      deal_damage,
+      check_for_death
+   ]
+   // can also use `chain![...]` for slightly more compact shortcut
+   .chain() 
+   .in_set(GameSet::Combat)
+)
+
+.add_systems(systems![run, jump].after(InputSet::ReadInput))
+```
+
+- looks OK (some boilerplate)
+- no limit to number of elements
+- relies on regular, declarative macro magic
+
+## Future possibilities
+
+There are a few proposals that were directly split from this RFC:
+
+1. Unnecessary dependency removal ([RFC #47](https://github.com/bevyengine/rfcs/pull/47)).
+2. Configurable system set atomicity ([RFC #46](https://github.com/bevyengine/rfcs/pull/46)).
+3. Automatic (opt-in) insertion of command application and state transition systems (see discussion in [RFC #34](https://github.com/bevyengine/rfcs/pull/34)).
+4. "Command-flushed" ordering constraints (dependencies that will ensure the pair is separated by command application).
+
+There are also lots of related but non-urgent areas for follow-up research:
+
+1. Improve tools for reporting and resolving system execution order ambiguities.
+2. Revisit plugin definition and plugin configuration.
+3. Cloning registered systems. (This probably warrants its own RFC, but it fits with the multiple worlds discussion.)
+4. Add [manual dependency graph construction](https://github.com/bevyengine/bevy/pull/2381) methods.
+5. Seeded, fully deterministic system execution orders: useful for debugging system order bugs and optimization work.
+6. Move command queues into the `World` so that exclusive systems can drain them directly (e.g. in some user-defined order).
+7. Compose conditions via arbitrary boolean expressions.
+8. Support automatic insertion and removal of systems to reduce schedule clutter and support other forms of one-off logic.
+9. Run schedules without `&mut World`, inferring access based on the systems inside.
+
+## Appendix
+
+These implementation ideas/details are here to avoid cluttering the RFC.
+
+### `apply_system_buffers`

```

Comment:
> I think more granular foo.after_commands(bar) scheduling (with commands being applied as late as possible) is the right long term bet.

Definitely agree, but this complexity is worth splitting off.

Reply:
Sure, `foo.after_commands(bar)` would be nice, but we don't know how to do that.

AFAIK none of us has a solution to "apply commands at a minimal number of points based on special granular dependencies." Seems like a funky optimal graph partitioning problem.

This RFC should at least make such a thing easier to implement tho, since there are no more loops to deal with.

Reply:
Cool cool. Sounds like this conversation is resolved!